File: Java/src/net/kenyang/algorithm/LinkedListCycle.java
package net.kenyang.algorithm;


public class LinkedListCycle {
    class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }
    

    public boolean hasCycle(ListNode head) {

        ListNode slowNode = head;
        ListNode fastNode = head;
        if (head !=null && head.next != null) {
            fastNode = head.next;
        } else {
            return false;
        }
                
        while (slowNode !=null && fastNode!=null){
            if (slowNode.val == fastNode.val ) return true;
            
            slowNode = slowNode.next;
            fastNode = fastNode.next;
            if (fastNode!=null){
                fastNode = fastNode.next;
            }
        }
        
        return false;
    }
}

File: Python/linked-list-cycle.py
# Time:  O(n)
# Space: O(1)
#
# Given a linked list, determine if it has a cycle in it.
# 
# Follow up:
# Can you solve it without using extra space?
#

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        fast, slow = head, head
        while fast and fast.next:
            fast, slow = fast.next.next, slow.next
            if fast is slow:
                return True
        return False

if __name__ == "__main__":
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = head.next
    print Solution().hasCycle(head)